src/libasr/codegen/asr_to_llvm.cpp

```diff
@@ -4651,14 +4651,58 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
         llvm::Value *left_val = tmp;
         this->visit_expr_wrapper(x.m_right, true);
         llvm::Value *right_val = tmp;
-        LCOMPILERS_ASSERT(ASRUtils::is_logical(*x.m_type))
+        llvm::Value *zero, *cond;
+        llvm::AllocaInst *result;
+        if (ASRUtils::is_integer(*x.m_type)) {
+            int a_kind = down_cast<ASR::Integer_t>(x.m_type)->m_kind;
+            int init_value_bits = 8*a_kind;
+            zero = llvm::ConstantInt::get(context,
+                            llvm::APInt(init_value_bits, 0));
+            cond = builder->CreateICmpEQ(left_val, zero);
+            result = builder->CreateAlloca(getIntType(a_kind), nullptr);
+        } else if (ASRUtils::is_real(*x.m_type)) {
+            int a_kind = down_cast<ASR::Real_t>(x.m_type)->m_kind;
+            int init_value_bits = 8*a_kind;
+            if (init_value_bits == 32) {
+                zero = llvm::ConstantFP::get(context,
+                                    llvm::APFloat((float)0));
+            } else {
+                zero = llvm::ConstantFP::get(context,
+                                    llvm::APFloat((double)0));
+            }
+            result = builder->CreateAlloca(getFPType(a_kind), nullptr);
+            cond = builder->CreateFCmpUEQ(left_val, zero);
+        } else if (ASRUtils::is_character(*x.m_type)) {
+            zero = llvm::Constant::getNullValue(character_type);
+            cond = lfortran_str_cmp(left_val, zero, "_lpython_str_compare_eq");
+            result = builder->CreateAlloca(character_type, nullptr);
+        } else if (ASRUtils::is_logical(*x.m_type)) {
+            zero = llvm::ConstantInt::get(context,
+                            llvm::APInt(1, 0));
+            cond = builder->CreateICmpEQ(left_val, zero);
+            result = builder->CreateAlloca(llvm::Type::getInt1Ty(context), nullptr);
```

Comment:
I think this operation should not be done at the backend level. A backend should exactly follow what the ASR says. In this case, it should always return a logical result. This logical result can/should later be casted to the desired type.

Reply:
Opened an issue here https://github.com/lfortran/lfortran/issues/3620.

Reply:
Yes, this should be explicitly done in ASR, not in the backend.